php/Block.php

```diff
@@ -63,66 +67,205 @@ public function register_block() {
+			 */
+			$posts = $this->get_tag_foo_cat_baz_posts();
+
+			// If the current post ID is part of the cached posts, remove it.
+			if ( isset ( $posts[ $current_post_id ] ) ) {
+				unset( $posts[ $current_post_id ] );
+			}
+
+			$posts = array_slice( $posts, 0, 5, true ); // Make sure the array contains atmost 5 posts.
+
+			if ( ! empty( $posts ) ) :
+				$post_count = count( $posts );
+
+				printf(
+					'<h2>%s</h2>',
+					esc_html( sprintf(
+					/* translators: %d: The number of posts with tag foo and category baz. */
+						_n(
+							'%d post with the tag of foo and the category of baz.',
+							'%d posts with the tag of foo and the category of baz.',
+							$post_count,
+							'site-counts'
+						),
+						$post_count
+					) )
+				); ?>
+
+				<ul>
+					<?php
+
+					foreach ( $posts as $post_title ) :
+						?>
+						<li><?php echo esc_html( $post_title ); ?></li>
+						<?php
+					endforeach;
+
+					?>
+				</ul>
+			<?php endif;
+			?>
+		</div>
+		<?php
+
+		return ob_get_clean();
+	}
+
+	/**
+	 * Retrieve post titles of posts with foo tag and baz category
+	 *
+	 * @param bool $force_refresh Whether to force the cache to be refreshed. Default false.
+	 *
+	 * @param int $post_count The number of posts to retrieve. Default 6.
+	 *
+	 * @return array Array of post titles matched to ids of posts containing foo tag and baz category
+	 */
+	public function get_tag_foo_cat_baz_posts( $force_refresh = false, $post_count = 6 ) {
+
+		// Check for site-counts_tag_foo_cat_baz_posts key in the 'block_posts' group.
+		$posts = wp_cache_get( 'site-counts_tag_foo_cat_baz_posts', 'block_posts' );
+
+		// If nothing is found or refresh is set to true, build the array.
+		if ( true === $force_refresh || false === $posts ) {
+			$posts = [];
+
+			$posts_query = new WP_Query( array(
+				'post_type'      => [ 'post', 'page' ],
+				'post_status'    => 'any',
+				'no_found_rows'  => true,
+				'posts_per_page' => $post_count,
+				'date_query'     => array(
 					array(
-						'hour'      => 9,
-						'compare'   => '>=',
+						'hour'    => 9,
+						'compare' => '>=',
 					),
 					array(
-						'hour' => 17,
-						'compare'=> '<=',
+						'hour'    => 17,
+						'compare' => '<=',
 					),
 				),
-                'tag'  => 'foo',
-                'category_name'  => 'baz',
-				  'post__not_in' => [ get_the_ID() ],
-			));
-
-			if ( $query->have_posts() ) :
-				?>
-				 <h2>5 posts with the tag of foo and the category of baz</h2>
-                <ul>
-                <?php
-
-                 foreach ( array_slice( $query->posts, 0, 5 ) as $post ) :
-                    ?><li><?php echo $post->post_title ?></li><?php
-				endforeach;
-			endif;
-		 	?>
-			</ul>
-		</div>
-		<?php
+				'tax_query'      => array(
+					'relation' => 'AND',
+					array(
+						'taxonomy'         => 'category',
+						'field'            => 'slug',
+						'include_children' => false,
```

Comment:
As of WordPress 4.4, terms have been split, which means that this now adds 'include_children' => true to almost all taxonomy queries. This can have a very significant performance impact on code that was performant previously and in some cases, queries that used to return almost instantly will time out. Thus the specification of
'include_children' => false.

 where the wanted behavior is to grab all posts that are either the parent or a child term this can be replaced by only querying for the parent term and enforcing inside a save_post() hook that if a child term is added then its parent term is also added. A one time WP-CLI command might be needed to ensure previous data integrity.